File: ios/LookSeeProto/Services/AuthService.swift
//
//  AuthService.swift
//  LookSeeProto
//
//  Created by Sheenan Ahsan on 2/25/26.
//
import Foundation
import Amplify
import AWSCognitoAuthPlugin
import AWSPluginsCore

class AuthService {

   static let shared = AuthService()

   private init() {}
   

   // SIGN UP
    func signUp(
        username _: String,
        password: String,
        email: String,
        group _: String = "authenticated-users"
    ) async throws -> AuthSignUpResult {
        let normalizedEmail = email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

        let options = AuthSignUpRequest.Options(
            userAttributes: [
                AuthUserAttribute(.email, value: normalizedEmail),
                AuthUserAttribute(
                    AuthUserAttributeKey(rawValue: "custom:group"),
                    value: "authenticated-users"
                )
            ]
        )
        return try await Amplify.Auth.signUp(
            username: normalizedEmail,
            password: password,
            options: options
        )
    }

   // SIGN IN
   func signIn(username: String, password: String) async throws -> AuthSignInResult {
       let normalizedUsername = username.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
       return try await Amplify.Auth.signIn(
           username: normalizedUsername,
           password: password
       )
   }

   // CONFIRM CODE
   func confirm(username: String, code: String) async throws -> AuthSignUpResult {
       return try await Amplify.Auth.confirmSignUp(
           for: username,
           confirmationCode: code
       )
   }

   // SIGN OUT
    func signOut() async {
        _ = await Amplify.Auth.signOut()
    }
    
    func fetchIdToken() async throws -> String {
        let session = try await Amplify.Auth.fetchAuthSession()

        guard let cognitoTokenProvider = session as? AuthCognitoTokensProvider else {
            throw AuthError.service(
                "Could not read Cognito session",
                "Session did not provide Cognito tokens",
                nil
            )
        }

        let tokens = try cognitoTokenProvider.getCognitoTokens().get()
        return tokens.idToken
    }

    // FETCH VERIFIED EMAIL — straight from Cognito's user attributes,
    // not whatever the app happens to have cached locally
    func fetchVerifiedEmail() async throws -> String? {
        let attributes = try await Amplify.Auth.fetchUserAttributes()
        return attributes.first(where: { $0.key == .email })?.value
    }

}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/services/AuthService.kt
package looksee.angelll.com.services

import com.amplifyframework.auth.AuthUserAttributeKey
import com.amplifyframework.auth.cognito.AWSCognitoAuthSession
import com.amplifyframework.auth.options.AuthSignUpOptions
import com.amplifyframework.auth.result.AuthSignInResult
import com.amplifyframework.auth.result.AuthSignUpResult
import com.amplifyframework.kotlin.core.Amplify

object AuthService {

    // SIGN UP
    suspend fun signUp(usernameInput: String, passwordInput: String, emailInput: String, groupInput: String): AuthSignUpResult {
        val options = AuthSignUpOptions.builder()
            .userAttribute(AuthUserAttributeKey.email(), emailInput)
            .userAttribute(AuthUserAttributeKey.custom("group"), groupInput) // store desired group
            .build()

        return Amplify.Auth.signUp(
            username = usernameInput, // 🚀 FIX: Now uses the username parameter!
            password = passwordInput,
            options = options
        )
    }

    // SIGN IN
    suspend fun signIn(usernameInput: String, passwordInput: String): AuthSignInResult {
        return Amplify.Auth.signIn(
            username = usernameInput,
            password = passwordInput
        )
    }

    // CONFIRM CODE
    suspend fun confirm(usernameInput: String, code: String): AuthSignUpResult {
        return Amplify.Auth.confirmSignUp(
            username = usernameInput,
            confirmationCode = code
        )
    }

    // SIGN OUT
    suspend fun signOut() {
        try {
            Amplify.Auth.signOut()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    // FETCH ID TOKEN
    suspend fun fetchIdToken(): String {
        val session = Amplify.Auth.fetchAuthSession() as? AWSCognitoAuthSession
            ?: throw Exception("Could not read Cognito session. Session did not provide Cognito tokens.")

        return session.userPoolTokensResult.value?.idToken
            ?: throw Exception("Could not retrieve ID token from session.")
    }

    // FETCH VERIFIED EMAIL
    suspend fun fetchVerifiedEmail(): String? {
        return try {
            val attributes = Amplify.Auth.fetchUserAttributes()
            attributes.find { it.key.keyString == "email" }?.value
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}